lib/AST/CompilerPlugin.cpp

```diff
@@ -0,0 +1,274 @@
+                    }
+                    exMetatype->addChild(type, D);
+                  }
+                  type->addChild(exMetatype, D);
+                }
+                typeList->addChild(type, D);
+              }
+              boundGenric->addChild(type, D);
+              boundGenric->addChild(typeList, D);
+            }
+            type->addChild(boundGenric, D);
+          }
+          variable->addChild(module, D);
+          variable->addChild(identifier, D);
+          variable->addChild(type, D);
+        }
+        getter->addChild(variable, D);
+      }
+      global->addChild(getter, D);
+    }
+    auto mangleResult = mangleNode(global);
+    assert(mangleResult.isSuccess());
+    name = mangleResult.result();
+  }
+  return ctx.getAddressOfSymbol(name, &library);
+}
+
+void ASTContext::loadCompilerPlugins() {
+  for (StringRef path : SearchPathOpts.getCompilerPluginLibraryPaths()) {
+    std::string errorMsg;
+    auto lib = llvm::sys::DynamicLibrary::getPermanentLibrary(
+        path.data(), &errorMsg);
+    if (!lib.isValid()) {
+      Diags.diagnose(SourceLoc(), diag::compiler_plugin_not_loaded, path,
+                     errorMsg);
+      continue;
+    }
+    auto moduleName = llvm::sys::path::filename(path);
+    #if !defined(_WIN32)
+    moduleName.consume_front("lib");
+    #endif
+    moduleName.consume_back(LTDL_SHLIB_EXT);
+    auto *getter = getMacroRegistrationPropertyGetter(lib, moduleName, *this);
+    if (!getter) {
+      Diags.diagnose(SourceLoc(),
+                     diag::compiler_plugin_missing_macro_declaration,
+                     moduleName, path, ALLMACROS_PROPERTY_NAME);
+      continue;
+    }
+    // Note: We don't currently have a way to poke at the contents of a Swift
+    // array `[Any.Type]` from C++. But this should not be an issue for release
+    // toolchains where user-defined macros will be used.
+    #if SWIFT_SWIFT_PARSER
+    auto types = swift_ASTGen_getMacroTypes(getter).array();
+    for (const void *type : types) {
+      CompilerPlugin plugin(type, lib, *this);
+      auto name = plugin.getName();
+      LoadedPlugins.try_emplace(name, std::move(plugin));
+    }
+    free(const_cast<void *>((const void *)types.data()));
+    #endif
+  }
+}
+
+using WitnessTableLookupFn = const void *(const void *type,
+                                          const void *protocol);
+
+#if SWIFT_SWIFT_PARSER
+extern "C" WitnessTableLookupFn swift_conformsToProtocol;
+#endif
+
+CompilerPlugin::CompilerPlugin(const void *metadata,
+                               llvm::sys::DynamicLibrary parentLibrary,
+                               ASTContext &ctx)
+   : metadata(metadata), parentLibrary(parentLibrary)
+{
+#if !SWIFT_SWIFT_PARSER
+  auto *swift_conformsToProtocol = reinterpret_cast<WitnessTableLookupFn *>(
+      ctx.getAddressOfSymbol("swift_conformsToProtocol"));
+#endif
+  void *protocolDescriptor = ctx.getAddressOfSymbol(
+      "$s22_CompilerPluginSupport01_aB0Mp");
```

Comment:
This string constant could benefit from being pulled out into a named constant like `COMPILER_PLUGIN_SUPPORT_PROTOCOL_DESCRIPTOR`.